src/computation/1_computation_parallel.cpp

```diff
@@ -327,3 +241,462 @@ void ComputationParallel::applyBoundaryValuesTop()
+
+            MPI_Recv(&other_ghost_to_v_right, 
+                     discretization_->nCells()[1]+4,
+                     MPI_DOUBLE,
+                     partitioning_.ownRankNo() - 1,
+                     )
+            MPI_Send(&v_to_other_ghost_right, 
+                     discretization_->nCells()[1]+4,
+                     MPI_DOUBLE,
+                     partitioning_.ownRankNo() - 1,
+                     )
+
+            for (int j = -2; j < discretization_->nCells()[1] +2; j++)
+            {
+                discretization_->v(discretization_->nCells()[1], j) = other_ghost_to_v_right[j+2]; // v_{n}
+            }
+        }
+        if (partitioning_.ownPartitionContainsLeftBoundary())
+        {
+            applyBoundaryValuesLeft();
+        } else 
+        {
+            // u
+            std::array<double, discretization_->nCells()[1]+4> u_to_other_ghost_left;
+            for (int j = -2; j < discretization_->nCells()[1] +2; j++)
+            {
+                u_to_other_ghost_left[j+2] = u(0, j); // u_{0}
+            }
+            std::array<double, discretization_->nCells()[1]+4> other_ghost_to_u_left;
+
+            MPI_Recv(&other_ghost_to_u_left, 
+                     discretization_->nCells()[1]+4,
+                     MPI_DOUBLE,
+                     partitioning_.ownRankNo() - 1,
+                     )
+            MPI_Send(&u_to_other_ghost_left, 
+                     discretization_->nCells()[1]+4,
+                     MPI_DOUBLE,
+                     partitioning_.ownRankNo() - 1,
+                     )
+
+            for (int j = -2; j < discretization_->nCells()[1] +2; j++)
+            {
+                discretization_->u(-2, j) = other_ghost_to_u_left[j+2]; // u_{-2}
+            }
+
+            // v
+            std::array<double, discretization_->nCells()[1]+4> v_to_other_ghost_left;
+            for (int j = -2; j < discretization_->nCells()[1] +2; j++)
+            {
+                v_to_other_ghost_left[j+2] = v(0, j); // v_{0}
+            }
+            std::array<double, discretization_->nCells()[1]+4> other_ghost_to_v_left;
+
+            MPI_Recv(&other_ghost_to_v_left, 
+                     discretization_->nCells()[1]+4,
+                     MPI_DOUBLE,
+                     partitioning_.ownRankNo() - 1,
+                     )
+            MPI_Send(&v_to_other_ghost_left, 
+                     discretization_->nCells()[1]+4,
+                     MPI_DOUBLE,
+                     partitioning_.ownRankNo() - 1,
+                     )
+
+            for (int j = -2; j < discretization_->nCells()[1] +2; j++)
+            {
+                discretization_->v(-1, j) = other_ghost_to_v_left[j+2]; // v_{-1}
+            }
+        }
+    }
+}
+
+ComputationParallel::uvExchangeVertical()
+{
+    // the +4 in nCells+4 is from the consistent number of halo-cells in the staggered grid
+
+    // the even processes: send top, receive top, send bottom, receive bottom
```

Comment:
even and uneven are the sam at the moment, should be changed

Reply:
Where? I don't think so